File: tools/domain_e2e/report.py
"""Console + Markdown report for domain e2e runs."""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Optional


@dataclass
class StepResult:
    name: str
    ok: bool
    duration_ms: float
    status_code: Optional[int] = None
    operation: Optional[str] = None
    instance_id: Optional[int] = None
    instance_status: Optional[str] = None
    last_error: Optional[str] = None
    captured: dict[str, Any] = field(default_factory=dict)
    errors: list[str] = field(default_factory=list)
    skipped: bool = False


@dataclass
class ScenarioResult:
    name: str
    path: str
    ok: bool
    duration_ms: float
    steps: list[StepResult] = field(default_factory=list)
    error: Optional[str] = None


def render_markdown(results: list[ScenarioResult], *, title: str = "Domain E2E Report") -> str:
    now = datetime.now(timezone.utc).strftime("%Y-%m-%d %H:%M:%S UTC")
    total = len(results)
    passed = sum(1 for r in results if r.ok)
    failed = total - passed
    lines = [
        f"# {title}",
        "",
        f"Generated: {now}",
        "",
        f"**Summary:** {passed}/{total} passed, {failed} failed",
        "",
    ]
    for sc in results:
        mark = "PASS" if sc.ok else "FAIL"
        lines.append(f"## [{mark}] {sc.name}")
        lines.append("")
        lines.append(f"- file: `{sc.path}`")
        lines.append(f"- duration: {sc.duration_ms:.0f} ms")
        if sc.error:
            lines.append(f"- error: `{sc.error}`")
        lines.append("")
        if sc.steps:
            lines.append("| Step | Result | HTTP | Instance | ms | Notes |")
            lines.append("|------|--------|------|----------|----|-------|")
            for st in sc.steps:
                if st.skipped:
                    res = "SKIP"
                elif st.ok:
                    res = "PASS"
                else:
                    res = "FAIL"
                http = str(st.status_code) if st.status_code is not None else "-"
                inst = st.instance_status or "-"
                if st.instance_id is not None:
                    inst = f"{inst} (#{st.instance_id})"
                notes = "; ".join(st.errors) if st.errors else ""
                if st.last_error and st.last_error not in notes:
                    notes = (notes + "; " if notes else "") + st.last_error
                if st.captured:
                    cap = ", ".join(f"{k}={v!r}" for k, v in st.captured.items())
                    notes = (notes + "; " if notes else "") + f"capture: {cap}"
                notes = notes.replace("|", "\\|")
                lines.append(
                    f"| {st.name} | {res} | {http} | {inst} | {st.duration_ms:.0f} | {notes} |"
                )
            lines.append("")
    return "\n".join(lines) + "\n"


def print_console(results: list[ScenarioResult]) -> None:
    total = len(results)
    passed = sum(1 for r in results if r.ok)
    print()
    print(f"=== Domain E2E: {passed}/{total} passed ===")
    for sc in results:
        mark = "PASS" if sc.ok else "FAIL"
        print(f"[{mark}] {sc.name} ({sc.duration_ms:.0f} ms)")
        if sc.error:
            print(f"  ERROR: {sc.error}")
        for st in sc.steps:
            if st.skipped:
                tag = "SKIP"
            elif st.ok:
                tag = "PASS"
            else:
                tag = "FAIL"
            extra = ""
            if st.errors:
                extra = " — " + "; ".join(st.errors)
            elif st.instance_status:
                extra = f" — instance={st.instance_status}"
            print(f"  [{tag}] {st.name}{extra}")
    print()


def write_report(path: Path, results: list[ScenarioResult]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(render_markdown(results), encoding="utf-8")

File: frontend/lib/e2e/report.ts
export type StepResult = {
  name: string;
  ok: boolean;
  durationMs: number;
  statusCode?: number | null;
  operation?: string | null;
  instanceId?: number | null;
  instanceStatus?: string | null;
  lastError?: string | null;
  captured: Record<string, unknown>;
  errors: string[];
  skipped?: boolean;
};

export type ScenarioResult = {
  name: string;
  source: string;
  ok: boolean;
  durationMs: number;
  steps: StepResult[];
  error?: string | null;
};

export function renderMarkdown(
  results: ScenarioResult[],
  title = "Domain E2E Report",
): string {
  const now = new Date().toISOString().replace("T", " ").replace(/\.\d+Z$/, " UTC");
  const total = results.length;
  const passed = results.filter((r) => r.ok).length;
  const failed = total - passed;
  const lines: string[] = [
    `# ${title}`,
    "",
    `Generated: ${now}`,
    "",
    `**Summary:** ${passed}/${total} passed, ${failed} failed`,
    "",
  ];

  for (const sc of results) {
    const mark = sc.ok ? "PASS" : "FAIL";
    lines.push(`## [${mark}] ${sc.name}`, "");
    lines.push(`- source: \`${sc.source}\``);
    lines.push(`- duration: ${sc.durationMs.toFixed(0)} ms`);
    if (sc.error) lines.push(`- error: \`${sc.error}\``);
    lines.push("");

    if (sc.steps.length) {
      lines.push("| Step | Result | HTTP | Instance | ms | Notes |");
      lines.push("|------|--------|------|----------|----|-------|");
      for (const st of sc.steps) {
        let res = "FAIL";
        if (st.skipped) res = "SKIP";
        else if (st.ok) res = "PASS";
        const http = st.statusCode != null ? String(st.statusCode) : "-";
        let inst = st.instanceStatus || "-";
        if (st.instanceId != null) inst = `${inst} (#${st.instanceId})`;
        let notes = st.errors.join("; ");
        if (st.lastError && !notes.includes(st.lastError)) {
          notes = notes ? `${notes}; ${st.lastError}` : st.lastError;
        }
        if (Object.keys(st.captured).length) {
          const cap = Object.entries(st.captured)
            .map(([k, v]) => `${k}=${JSON.stringify(v)}`)
            .join(", ");
          notes = notes ? `${notes}; capture: ${cap}` : `capture: ${cap}`;
        }
        notes = notes.replace(/\|/g, "\\|");
        lines.push(
          `| ${st.name} | ${res} | ${http} | ${inst} | ${st.durationMs.toFixed(0)} | ${notes} |`,
        );
      }
      lines.push("");
    }
  }
  return `${lines.join("\n")}\n`;
}

export function downloadMarkdown(content: string, filename?: string) {
  const stamp = new Date()
    .toISOString()
    .replace(/[-:TZ.]/g, "")
    .slice(0, 14);
  const name = filename || `e2e_${stamp}.md`;
  const blob = new Blob([content], { type: "text/markdown;charset=utf-8" });
  const url = URL.createObjectURL(blob);
  const a = document.createElement("a");
  a.href = url;
  a.download = name;
  a.click();
  URL.revokeObjectURL(url);
}
